packages/lit-html/src/directives/async-replace.ts

```diff
@@ -0,0 +1,115 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {ChildPart, noChange} from '../lit-html.js';
+import {
+  directive,
+  DirectiveParameters,
+  PartInfo,
+  PartType,
+} from '../directive.js';
+import {AsyncDirective} from '../async-directive.js';
+
+type Mapper<T> = (v: T, index?: number) => unknown;
+
+class AsyncReplaceDirective extends AsyncDirective {
+  value?: AsyncIterable<unknown>;
+  reconnectResolver?: () => void;
+  reconnectPromise?: Promise<void>;
+
+  constructor(partInfo: PartInfo) {
+    super(partInfo);
+    if (partInfo.type !== PartType.CHILD) {
+      throw new Error('asyncAppend can only be used in child expressions');
+    }
+  }
+
+  render<T>(_value: AsyncIterable<T>, _mapper?: Mapper<T>) {
+    return noChange;
+  }
+
+  update(_part: ChildPart, [value, mapper]: DirectiveParameters<this>) {
+    // If we've already set up this particular iterable, we don't need
+    // to do anything.
+    if (value === this.value) {
+      return;
+    }
+    this.value = value;
+    this.__iterate(mapper);
+    return noChange;
+  }
+
+  // Separate function to avoid an iffe in update; update can't be async
+  // because its return value must be `noChange`
+  private async __iterate(mapper?: Mapper<unknown>) {
+    let i = 0;
+    const {value} = this;
+    for await (let v of value!) {
+      // Check to make sure that value is the still the current value of
+      // the part, and if not bail because a new value owns this part
+      if (this.value !== value) {
+        break;
+      }
+
+      // If we were disconnected, pause until reconnected
+      if (this.reconnectPromise) {
+        await this.reconnectPromise;
+      }
+
+      // As a convenience, because functional-programming-style
+      // transforms of iterables and async iterables requires a library,
+      // we accept a mapper function. This is especially convenient for
+      // rendering a template for each item.
+      if (mapper !== undefined) {
+        v = mapper(v, i);
+      }
+
+      this.setValue(v);
```

Comment:
@justinfagnani The original 1.x implementation created a nested child part here, but it was sorta unclear why. I'm just using the parent part for replace. 🤷

Reply:
I think this change makes asyncReplace compatible with attribute/property parts, yeah?

Maybe document/test that? It could be useful.

Reply:
Yeah good point. Done.

I thought about ways it might make sense to have a switch on part type in `asyncAppend` to keep the current behavior for `ChildPart` and do something different for the other part types, but it doesn't really make too much sense I think. You could either concat as string (maybe good for attributes, but then you'd probably want a joiner), or else spread into a new array each time (maybe good for properties), but not a slam dunk and not worth going there now.